### ISSUE TITLE
Timeline: How to draw lines/arrows to connect items?

### ISSUE THREAD

**Issue by Rikk** [almende/vis#1699]
Hello, I've started looking at vis.js recently. Its graphing capabilities look very promising!
Now, I could not found this in any example. May I've missed it?
I need to create a timeline graph with some networking, or in other words, an way to express relation between items in a timeline.
For example:
![timeline-network](https://cloud.githubusercontent.com/assets/2916485/13415250/0fd1c3f2-df38-11e5-9c3f-84b862a33f5f.png)
Is this possible? What is needed to implement this?

**Comment by josdejong** [comment]
This is simply not possible right now. Implementing it is possible but a lot of work. We don't have plans for that ourselves but if anyone wants to pick this up would be interesting.

**Comment by Rikk** [comment]
Does vis.js provide any way to draw arrows in timeline? Or only in network mode? How are they rendered in network mode?

**Comment by AlexDM0** [comment]
Hi,

Network is completely separated from the timeline. It's rendered in canvas. I do not think any attempt can be made to merge this into timeline. 

Cheers

**Comment by josdejong** [comment]
Indeed. Network uses HTML5 Canvas, Timeline uses HTML DOM, and Graph2d uses HTML DOM and SVG. 

If we extend the Timeline with these sort of arrows, I think the best approach is to add an SVG layer to the Timeline where you can render arrows and stuff.

**Comment by Rikk** [comment]
> If we extend the Timeline with these sort of arrows, I think the best approach is to add an SVG layer to the Timeline where you can render arrows and stuff

This would be great.

**Comment by mojoaxel** [comment]
There is a [jquery plugin for gantt charts](http://mbielanczuk.com/jquery-gantt/):

![image](https://cloud.githubusercontent.com/assets/600565/18045261/474f8d68-6dd2-11e6-9e61-fd61e063aed3.png)

Maybe you can use this instead?

**Comment by yotamberk** [comment]
@Rikk Did the plumpjs solve the issue? Can we close this issue?

**Comment by mojoaxel** [comment]
@yotamberk I would like to keep this feature-request open until it is implemented.

**Comment by yotamberk** [comment]
@mojoaxel Sure. I think this issue is extremely difficult to implement... I think it should be implemented with an external library (like plumbjs). I'm considering to pick this feature and see what it actually means. 
Although I feel it will be much more complicated than we can handle...

**Comment by Rikk** [comment]
I think of jsPlumb as a workaround. It's a big lib, too complex, docs are confusing and most of its functions were not used. I couldn't figure out the correct way not to overlap or crossover items (it would need some logic I didn't think about); I used other visual strategies to make this supportable. Whatever, it's root use is to connect html elements on the page; these elements are the items generated by vis.js.

Maybe other than using svg, it could be implemented using invisible elements and css tricks. Possibly svg is easier and performs better, but I don't know how to program svg. It would need to recalculate source and destination points and redraw as timeline redraws and update a `<path>` inside the `<svg>`.

The gantt-charts plugins mentioned looks too restrictive, eg. only one item per group, development stopped, any kind of implementation would require deep modifications, and so on. Not good bet.

**Comment by mojoaxel** [comment]
> Maybe other than using svg, it could be implemented using invisible elements and css tricks. Possibly svg is easier and performs better, but I don't know how to program svg

Have a look into [d3.js](https://d3js.org/).

There are also more [(commercial) players](https://dhtmlx.com/docs/products/dhtmlxGantt/) out there:
![image](https://cloud.githubusercontent.com/assets/600565/19707000/4fcc88ea-9b15-11e6-98d6-9e533a2da542.png)

**Comment by corycowgill** [comment]
This would be very helpful as we can get 90% of what we want with Timeline but want to connect the items in the timeline with directional arrows similar to above.

**Comment by sihil** [comment]
Another +1 on having this sort of support. A hybrid timeline-network would be really useful on my current project. It would be interesting to either be able to express relationships in a timeline or be able to align items in a network so as to be able to create a form of timeline.

**Comment by frboyer** [comment]
I wanted to verify how hard it would be to integrate arrows in vis.js timeline and if it really requires "breaking changes".  I made a small example to show that it is possible to add an arrow without touching the source code of vis.js, thus it would not be that "breaking" to integrate some arrows.  I used an SVG layer to draw the arrow, and hijacked the "on changed" event to update the arrow with the view.

The source html/js is [here](https://gist.github.com/frboyer/c5484b23b401c33be8abaea766783402) and you can view it works [here](https://gistpreview.github.io/?c5484b23b401c33be8abaea766783402).

**Comment by zivo13** [comment]
Very interesting approach by frboyer, can it work for multiple items?

**Comment by zdenal** [comment]
Does anybody found some simple/clever solution for this? Could anybody help me (give me a clue) find some tool by which could achieve something like this?

<img width="1425" alt="Greg_Young_on_Twitter___Hey_whats_this_new_feature_in__eventstore…__" src="https://user-images.githubusercontent.com/143357/58729958-1090f780-83eb-11e9-9a48-9f4b71e24784.png">

**Comment by JimmyCheng** [comment]
Thanks @frboyer.  I have revised @frboyer 's gist and added multiple dependencies in the vis.js project
@zivo13 , @zdenal this may help you.

The source html/js is [here](https://gist.github.com/JimmyCheng/b5a7a48e51d0b5d62aee718e10039e16) and you can view it works [here](https://gistpreview.github.io/?b5a7a48e51d0b5d62aee718e10039e16).

**Comment by javdome** [comment]
@JimmyCheng it's amazing!!!
I have added the selectable and editable options to the timeline and it is even more amazing, since the items can be dragged.
See [here](https://gistpreview.github.io/?5c09b1cc4b21614255e8d426a6117e84)

**Comment by JimmyCheng** [comment]
The best solution to solve this issue is to use visjs + [leaderline](https://github.com/anseki/leader-line). I can provide an example later.
